### PR TITLE
Enable relu fusion with prepacked linear/conv.

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35705 Enable relu fusion with prepacked linear/conv.**

Summary:
Introduces a pass for relu fusion.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20746592](https://our.internmc.facebook.com/intern/diff/D20746592)

### CODE REVIEW

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 1]
```diff
@@ -81,8 +81,105 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
+          %dilation:int[], %groups:int, %output_min_max):
+        %output_min: float = prim::Constant[value=0.0]()
+        %output_max: float = prim::Constant[value=1.0]()
+        %packed_weight_bias = prepacked::conv2d_clamp_prepack(
+            %weight, %bias, %stride, %padding, %dilation, %groups,
+            %output_min, %output_max)
+        %r = prepacked::conv2d_clamp_run(%input, %packed_weight_bias)
+        return (%r) )";
+
+  if (isReluFusable(graph)) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp]: looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: > looks like this is checking for all `aten::relu` in the graph but the pattern is based on individual aten::relu?

Good point. Yes this checks for all relus.
> 
> I think when the input of relu is used once in the pattern it is only going to match the case when the input is used once as well, have you tried removing this check?

Oh I see. I need to remove this and add a test where output of conv is used elsewhere as well and check.

Comment (jerryzh168) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: btw, this is only true if the value is not input: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/subgraph_matcher.cpp.html#100

Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: Ah. Makes sense. Actually, I dont need to test. This makes sense. I was only trying to check that relu is not fused with previous node, if the output of the previous node is used elsewhere. If input to relu is input to the pattern graph then there is nothing relu can fuse into.

File: test/test_xnnpack_integration.py [hunk 2]
```diff
@@ -479,24 +479,52 @@ def __init__(self):
...
+        pattern_count_map = {"aten::relu": 2,
+                             "prepacked::conv2d_clamp_prepack": -1,
+                             "prepacked::conv2d_clamp_run": 1,
+                             "prepacked::linear_clamp_prepack": -1,
+                             "prepacked::linear_clamp_run": 1}
+        validate_transformed_module(M(), pattern_count_map, data_shape, True)
+        pattern_count_map["prepacked::conv2d_clamp_prepack"] = -1
+        pattern_count_map["prepacked::linear_clamp_prepack"] = -1
+        pattern_count_map["aten::relu"] = -1
+        validate_transformed_module(M(), pattern_count_map, data_shape, True, True)
```
Comment (dreiss) [inline test/test_xnnpack_integration.py]: See https://our.intern.facebook.com/intern/wiki/Facebook-code-smells/boolean-arguments/ .  I think using booleans is fine here, but pass the args by keyword instead of position so it's clear what they mean.

File: torch/csrc/jit/passes/graph_rewrite_helper.h [hunk 3]
```diff
@@ -17,6 +18,27 @@ c10::optional<IValue> getIValue(
...
+// runtime cost
+struct PatternInfo {
+  std::string pattern_string;
+  std::unique_ptr<Graph> pattern_graph;
+  std::unordered_map<std::string, Value*> vmap;
+
+  static PatternInfo parse_from_str(std::string pattern_string) {
+    PatternInfo rv{
+        std::move(pattern_string),
+        std::make_unique<Graph>(), decltype(vmap){}};
```
Comment (dreiss) [inline torch/csrc/jit/passes/graph_rewrite_helper.h]: formatting looks off.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 4]
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
+// be done inplace. However, if the input of the relu was not to be
+// overwritten then it should not be fused.
+void fuseReluWithPackedOps(std::shared_ptr<Graph>& graph) {
+
+  SubgraphRewriter rewriter;
+
+  std::string linear_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %output_min: float = prim::Constant[value=0.0]()
+        %output_max: float = prim::Constant[value=1.0]()
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp]: Shouldn't output_max be None for relu?

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 5]
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
 
+// Fusing Relu with Linear/Conv already implies that relu will
+// be done inplace. However, if the input of the relu was not to be
+// overwritten then it should not be fused.
+void fuseReluWithPackedOps(std::shared_ptr<Graph>& graph) {
+
+  SubgraphRewriter rewriter;
+
+  std::string linear_prepack_run_relu_fused = R"(
+    graph(%input, %weight, %bias, %output_min_max):
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp]: What is output_min_max?  Isn't this supposed to appear in the pattern below?

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 6]
```diff
@@ -86,6 +86,82 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
+  std::string linear_prepack_run_relu = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %packed_weight_bias = prepacked::linear_clamp_prepack(
+            %weight, %bias, %output_min_max, %output_min_max)
+        %linear_res = prepacked::linear_clamp_run(%input, %packed_weight_bias)
+        %res = aten::relu(%linear_res)
+        return (%res))";
+
+  rewriter.RegisterRewritePattern(linear_prepack_run_relu,
+      linear_prepack_run_relu_fused);
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp]: I'm confused.  Above, you said  "if the input of the relu was not to be overwritten then it should not be fused.", but isn't it being fused here?

**Review by dreiss — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: Thanks for the catch.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 5 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: Graph fuser will complain if the graph being replaced with has a different signature. Thats why I had to keep it here.
 also realized, I think we will need to do constant pooling after this pass, because for every graph being replaced we are introducing the constant of '0'.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 6 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp]: It is a stale comment. I removed some code and forgot to remove the comment. Thanks for catching it. Will remove the comment.

File: torch/csrc/jit/python/init.cpp [hunk 7]
```diff
@@ -488,6 +488,11 @@ void initJITBindings(PyObject* module) {
       .def(
           "_jit_pass_insert_prepacked_ops",
           [](script::Module& module) { return insertPrePackedOps(module); })
+      .def(
+          "_jit_pass_fuse_relu_w_prepacked_linear_conv",
+          [](script::Module& module) {
+            return fusePrePackedLinearConvWithRelu(module);
```
Comment (AshkanAliabadi) [inline torch/csrc/jit/python/init.cpp]: Do we also support different variations of ReLU6 and Hardtanh in this rule?

Comment (kimishpatel) [reply torch/csrc/jit/python/init.cpp]: I did not add that as part of this, but given the patters are just string. We can replace relu with hardtanh. So should be doable.

Comment (AshkanAliabadi) [reply torch/csrc/jit/python/init.cpp]: Can you please enable ReLU6 as well (if not already part of this PR.)  Reason being that it is used in MobileNetV2 and I am sure it won't be long before we receive a usecase in which that is causing a performance issue.

Comment (kimishpatel) [reply torch/csrc/jit/python/init.cpp]: Sure. Will do this in this PR itself. Note that ReLU6 in pytorch just maps to hardtanh.

File: test/test_xnnpack_integration.py [hunk 8]
```diff
@@ -199,10 +199,10 @@ def forward(self, x):
...
         paddings = (pad_h, pad_w)
         dilations = (dilation, dilation)
-        assume(height + 2 * paddings[0] >=
-               dilations[0] * (kernels[0] - 1) + 1)
-        assume(width + 2 * paddings[1] >=
-               dilations[1] * (kernels[1] - 1) + 1)
+        assume(height + 2 * paddings[0]
+               >= dilations[0] * (kernels[0] - 1) + 1)
+        assume(width + 2 * paddings[1]
+               >= dilations[1] * (kernels[1] - 1) + 1)
```
Comment (dreiss) [inline test/test_xnnpack_integration.py:205]: Please avoid mixing unrelated formatting changes.

File: test/test_xnnpack_integration.py [hunk 9]
```diff
@@ -366,29 +366,30 @@ def forward(self, x):
...
             scripted_model.eval()
             input_data = torch.rand(data_shape)
             ref_result = scripted_model(input_data)
             torch._C._jit_pass_insert_prepacked_ops(scripted_model._c)
-            if (prepack_removal):
+            if fuse_clamping_ops or prepack_removal:
                 scripted_model._c = torch._C._freeze_module(scripted_model._c)
+            if fuse_clamping_ops:
+                torch._C._jit_pass_fuse_clamp_w_prepacked_linear_conv(scripted_model._c)
+            if (prepack_removal):
```
Comment (dreiss) [inline test/test_xnnpack_integration.py:379]: Don't need parens.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 10]
```diff
@@ -86,6 +87,147 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
+        %r = aten::relu(%conv2d_res)
+        return (%r) )";
+
+  rewriter.RegisterRewritePattern(
+      conv2d_prepack_run_relu, conv2d_prepack_run_relu_fused);
+
+  std::string linear_prepack_run_relu_inplace = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %packed_weight_bias = prepacked::linear_clamp_prepack(
+            %weight, %bias, %output_min_max, %output_min_max)
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:209]: Does this verify that output_min_max is None?  If not (for example, if it's 1), this would change semantics.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 11]
```diff
@@ -86,6 +87,147 @@ void insertPrePackedConv2dOp(std::shared_ptr<Graph>& graph) {
...
+        %r = prepacked::conv2d_clamp_run(%input, %packed_weight_bias)
+        return (%r) )";
+
+  std::string linear_prepack_run_relu = R"(
+    graph(%input, %weight, %bias, %output_min_max):
+        %packed_weight_bias = prepacked::linear_clamp_prepack(
+            %weight, %bias, %output_min_max, %output_min_max)
+        %linear_res = prepacked::linear_clamp_run(%input, %packed_weight_bias)
+        %res = aten::relu(%linear_res)
+        return (%res))";
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:188]: Does this verify that linear_res is not used elsewhere?  If not, this change is not semantics-preserving.

File: test/test_xnnpack_integration.py (hunk 8 above)
Comment (kimishpatel) [reply test/test_xnnpack_integration.py:205]: Sorry about this. I just ran autopep for formatting.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 10 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:209]: No this does not check for the values, it simply reroutes the values. Good catch. The place it fails is:
```
conv2d
hardtanh(-1, 1)
hardtanh(0, 6)
```
On pattern like the above one, if you transform conv2d to conv2d_clamp_prepack + conv2d_clamp_run, and then run clamping fusion pass on it you will get:
```
conv2d_clamp_prepack (min=-1, max=1)
conv2d_clamp_run
hardtanh(0, 6)
```
If you run the pass again on this. We will get pattern that actually gives you wrong result. I think we can fix this. Thanks for the catch.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 11 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:188]: I had a check earlier to do precisely this. However Jerry pointed out here, https://github.com/pytorch/pytorch/pull/35705#discussion_r400559498,  that this is already taken care of by subgraph rewriter.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 12]
```diff
@@ -93,15 +93,12 @@ bool bothMinMaxNone(
   const auto& match_vmap = match.values_map;
   // If value map has dummy_min_max in it, then it means
   // the pattern to be replaced was the one containing hardtanh.
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:95]: Comment is stale.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 13]
```diff
@@ -93,15 +93,12 @@ bool bothMinMaxNone(
...
-  // Otherwise the pattern is relu.
-  auto output_min_max =
-    graph_rewrite_helper::getIValue("output_min_max", match_vmap, vmap);
-  return (!output_min_max || output_min_max.value().isNone());
+  TORCH_CHECK(
+      vmap.find("dummy_min_max") != vmap.end(),
+      "Expected to find dummy_min_max Value in the subgraph to be replaced.");
+  auto dummy_min_max =
+    graph_rewrite_helper::getIValue("dummy_min_max", match_vmap, vmap);
+  return (!dummy_min_max || dummy_min_max.value().isNone());
```
Comment (dreiss) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:101]: Is `!dummy_min_max` still valid?

**Review by dreiss — APPROVED** [review]

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 13 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:101]: Just guarding against the possibility that instead of None Value, getIValue returns c10::nullopt.
